pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/dictionary/BaseOffHeapMutableDictionary.java

```diff
@@ -0,0 +1,411 @@
+ * Here is the pseudo-code for indexing an item or finding the dictionary ID of an item.
+ *
+ * index(item) {
+ *   foreach (iBuf: iBufList) {
+ *     hash value into a row for the buffer.
+ *     foreach (cell: rowOfCells) {
+ *       if (cell is not occupied) {
+ *         set it to dictId
+ *         return
+ *       } else if (item.equals(get(dictId)) {
+ *         // item already present in dictionary
+ *         return
+ *       }
+ *     }
+ *   }
+ *   oveflow.put(item, dictId)
+ *   if (overflow.size() > threshold) {
+ *     newSize = lastBufSize * expansionMultiple
+ *     newBuf = allocateDirect(newSize)
+ *     add newBuf to iBufList
+ *     newOverflow = new HashMap()
+ *     foreach (entry: overflow) {
+ *       hash entry.key() into a row for newBuf
+ *       foreach (cell : rowOfCells) {
+ *         if (cell empty) {
+ *           set cell to entry.value();
+ *         }
+ *       }
+ *       if (we did not enter value above) {
+ *         newOverflow.put(entry.key(), entry.value());
+ *       }
+ *     }
+ *   }
+ * }
+ *
+ * indexOf(item) {
+ *   foreach (iBuf: iBufList) {
+ *     hash value into a row for the buffer;
+ *     foreach (cell : rowOfCells) {
+ *       if (cell is not occupied) {
+ *         return NULL_VALUE_INDEX;
+ *       }
+ *       if (cell is occupied && item.equals(get(dictId))) {
+ *         return dictId;
+ *       }
+ *     }
+ *   }
+ *   if (overflow.contains(item)) {
+ *     return overflow.get(item);
+ *   }
+ *   return NULL_VALUE_INDEX;
+ * }
+ *
+ * The list of buffers and the overflow hash are maintained in a class (ValueToDictId) that is referenced via an
+ * AtomicReference. This ensures that readers always get either the new version of these objects or the old version,
+ * but not some inconsistent versions of these.
+ *
+ * It should be noted that this class assumes that there is one writer and multiple readers of the dictionary. It is
+ * NOT safe for a multiple writer scenario.
+ *
+ * TODO
+ * - It may be useful to implement a way to stop adding new items when the the number of buffers reaches a certain
+ *   threshold. In this case, we could close the realtime segment, and start a new one with bigger buffers.
+ */
+public abstract class BaseOffHeapMutableDictionary extends MutableDictionary {
+  // Percent of cardinality that we expect to hold in the overflow hash table.
+  private static final int OVERFLOW_THRESHOLD_PERCENT = 1;
+
+  // expansionMultiple setting as we add new buffers. A setting of 1 sets the new buffer to be
+  // the same size as the last one added. Setting of 2 allocates a buffer twice as big as the
+  // previous one. It is assumed that these values are powers of 2. It is a good idea to restrict
+  // these to 1 or 2, but not higher values. This array can be arbitrarily long. If the number of
+  // buffers exceeds the number of elements in the array, we use the last value.
+  private static final int[] EXPANSION_MULTIPLES = new int[] {1, 1, 2, 2, 2};
```

Comment:
The idea is to keep the implementation flexible at this point, because we don't have enough data right now. Having the expansion multiple be configurable helps us evaluate different expansion techniques in our benchmark.